exporter/awsemfexporter/pusher.go

```diff
@@ -0,0 +1,313 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awsemfexporter
+
+import (
+	"log"
+	"sort"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+)
+
+const (
+	//http://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/cloudwatch_limits_cwl.html
+	//In truncation logic, it assuming this constant value is larger than PerEventHeaderBytes + len(TruncatedSuffix)
+	MaxEventPayloadBytes = 1024 * 256 //256KB
+	// http://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html
+	MaxRequestEventCount   = 10000
+	PerEventHeaderBytes    = 26
+	MaxRequestPayloadBytes = 1024 * 1024 * 1
+
+	logEventChanBufferSize    = 10000 // 1 request can handle max 10000 log entries
+	minPusherIntervalInMillis = 200   // 5 TPS
+
+	logEventBatchPushChanBufferSize = 2 // processing part does not need to be blocked by the current put log event request
+	TruncatedSuffix                 = "[Truncated...]"
+)
+
+//Struct to present a log event.
+type LogEvent struct {
+	InputLogEvent *cloudwatchlogs.InputLogEvent
+	//The time which log generated.
+	LogGeneratedTime time.Time
+}
+
+//Calculate the log event payload bytes.
+func (logEvent *LogEvent) eventPayloadBytes() int {
+	return len(*logEvent.InputLogEvent.Message) + PerEventHeaderBytes
+}
+
+func (logEvent *LogEvent) truncateIfNeeded() bool {
+	if logEvent.eventPayloadBytes() > MaxEventPayloadBytes {
+		log.Printf("W! logpusher: the single log event size is %v, which is larger than the max event payload allowed %v. Truncate the log event.", logEvent.eventPayloadBytes(), MaxEventPayloadBytes)
```

Comment:
Oh can you make sure to use the zap logger throughout the PR? We don't use standard library logging in the collector

Reply:
will do. Thanks